aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/Orchestrator.java

```diff
@@ -142,24 +142,53 @@ public boolean isStopped() {
         return Mode.STOPPED.equals(currentMode.get());
     }
 
+    /**
+     * Start the orchestrator with the default timeout of 10 seconds.
+     * @return A completable that when subscribed to will attempt to start the orchestrator.
+     */
+    public synchronized Completable start() {
+        return start(NETWORK_OP_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+    }
+
     /**
      * Start performing sync operations between the local storage adapter
-     * and the remote GraphQL endpoint.
+     * and the remote GraphQL endpoint. The timeout parameters provided will be
+     * used to block execution of the current thread until lock acquisition has
+     * either failed or succeded, or the timeout is elapsed.
+     *
+     * @param opTimeout The desired timeout for the start operation.
+     * @param timeUnit The unit of time of the opTimeout parameter.
+     * @return A completable that when subscribed to will attempt to start the orchestrator.
      */
-    public void start() {
-        disposables.add(transitionCompletable()
-            .subscribeOn(startStopScheduler)
-            .doOnDispose(() -> LOG.debug("Orchestrator disposed a transition."))
-            .subscribe(
-                () -> {
-                    LOG.debug("Orchestrator completed a transition");
-                    if (isStarted()) {
-                        Amplify.Hub.publish(HubChannel.DATASTORE,
-                            HubEvent.create(DataStoreChannelEventName.READY));
-                    }
-                },
-                failure -> LOG.warn("Orchestrator failed to transition.")
-            ));
+    public synchronized Completable start(long opTimeout, TimeUnit timeUnit) {
+        if (tryAcquireStartStopLock(opTimeout, timeUnit)) {
+            return transitionCompletable()
+                .doOnSubscribe(subscriber -> {
+                    LOG.debug("Starting the orchestrator.");
+                })
+                .doOnComplete(() -> LOG.debug("Orchestrator started."))
+                .doOnDispose(() -> LOG.debug("Orchestrator disposed a transition."))
+                .doFinally(startStopSemaphore::release)
+                .subscribeOn(startStopScheduler);
+        } else {
+            return Completable.error(new DataStoreException("Unable to acquire orchestrator lock. " +
+                                                        "Transition currently in progress.",
+                                                        "Retry your operation."));
+        }
+    }
+
+    private boolean tryAcquireStartStopLock(long opTimeout, TimeUnit timeUnit) {
+        LOG.debug("Attempting to acquire lock. Permits available = " + startStopSemaphore.availablePermits());
```

Comment:
Considering that we are using this lock for mutual exclusion, isn't this log redundant?

Reply:
@raphkim I think your point is that this is a "yes/no" log, not a "0...k" log, right? Or in other words, the observation is that `availablePermits()` is either 0 or 1?

Reply:
Yep! It's either available or not haha

Reply:
It's easy enough to make that true or false instead.